langserver/diagnostics.go

```diff
@@ -18,18 +19,37 @@ import (
 
 type diagnostics map[string][]*lsp.Diagnostic // map of URI to diagnostics (for PublishDiagnosticParams)
 
+type diagnosticsCache struct {
+	cache diagnostics
+	lock  sync.Mutex
+}
+
+func (p *diagnosticsCache) update(fn func(diagnostics) diagnostics) {
```

Comment:
could you document this behaviour here? That we enforce that only one fn is running at a time.